src/mlmc/quantity.py

```diff
@@ -112,31 +112,34 @@ def size(self) -> int:
         """
         return self.qtype.size()
 
-    def get_cache_key(self, chunk_spec):
+    def get_cache_key(self, level_id=None, chunk_id=None, chunk_slice=None, n_samples=None):
         """
         Create cache key
-        :param chunk_spec: ChunkSpec instance
-        :return: tuple
         """
-        return (chunk_spec.level_id, chunk_spec.chunk_id, chunk_spec.n_samples, chunk_spec.chunk_size, id(self))  # redundant parentheses needed due to py36, py37
+        chunk_size = None
+        if chunk_slice is not None:
+            chunk_size = chunk_slice.stop - chunk_slice.start
+        return (level_id, chunk_id, chunk_size, n_samples, id(self))  # redundant parentheses needed due to py36, py37
 
     @cached(custom_key_maker=get_cache_key)
-    def samples(self, chunk_spec):
+    def samples(self, level_id=None, chunk_id=None, chunk_slice=None, n_samples=None):
         """
-        Yields list of sample chunks for individual levels.
+        Return list of sample chunks for individual levels.
         Possibly calls underlying quantities.
-        :param chunk_spec: ChunkSpec instance, it contains level_id, chunk_id and
-                                               it may contain n_samples - number of which we want to retrieve
+        :param level_id: int, level identifier
+        :param chunk_id: int, chunk identifier
+        :param chunk_slice: slice() object
+        :param n_samples: int, number of samples to retrieve
```

Comment:
`n_samples` should rather be parameter of Storage.chunks

Reply:
I find it too complicated for such a simple use of `n_samples`.

Reply:
No, it is fairly simple and significantly simplifies code on many places.

Reply:
ok